File: THCS/2022_2023/HoaBinh_2223/Bai2/BPAIRS.py
def digit_sum(n):
    """Tính tổng các chữ số của số nguyên n."""
    return sum(int(digit) for digit in str(n))

def count_beautiful_pairs(numbers):
    """Đếm số lượng cặp số đẹp trong danh sách numbers."""
    sum_count = {}
    beautiful_pairs = 0

    for number in numbers:
        s = digit_sum(number)  # Tính tổng chữ số
        if s in sum_count:
            beautiful_pairs += sum_count[s]  # Cộng dồn số cặp đẹp
            sum_count[s] += 1  # Tăng số lần xuất hiện của tổng chữ số
        else:
            sum_count[s] = 1  # Khởi tạo số lần xuất hiện của tổng chữ số

    return beautiful_pairs

# Nhập dữ liệu
n = int(input())
numbers = list(map(int, input().split()))

# Tính và xuất kết quả
result = count_beautiful_pairs(numbers)
print(result)

File: THCS/2022_2023/HoaBinh_2223/Bai2/BPAIRS.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

// Hàm tính tổng các chữ số của một số nguyên
int digitSum(int n) {
    int sum = 0;
    while (n > 0) {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

int main() {
    int n;
    cin >> n;

    vector<int> a(n);
    unordered_map<int, int> sumCount; // Lưu số lần xuất hiện của mỗi tổng chữ số
    long long beautifulPairs = 0; // Số lượng cặp số đẹp

    // Đọc dãy số
    for (int i = 0; i < n; i++) {
        cin >> a[i];
        int sum = digitSum(a[i]); // Tính tổng chữ số
        beautifulPairs += sumCount[sum]; // Thêm số lượng cặp số đẹp có thể được tạo ra
        sumCount[sum]++; // Tăng số lần xuất hiện của tổng chữ số
    }

    cout << beautifulPairs << endl; // Xuất kết quả
    return 0;
}
